src/pip/_internal/network/lazy_wheel.py

```diff
@@ -47,20 +58,60 @@ class LazyZipOverHTTP:
     """
 
     def __init__(
-        self, url: str, session: PipSession, chunk_size: int = CONTENT_CHUNK_SIZE
+        self, url: str, session: Session, chunk_size: int = CONTENT_CHUNK_SIZE
     ) -> None:
-        head = session.head(url, headers=HEADERS)
-        raise_for_status(head)
-        assert head.status_code == 200
+
+        # if CONTENT_CHUNK_SIZE is bigger than the file:
+        # In [8]: response.headers["Content-Range"]
+        # Out[8]: 'bytes 0-3133374/3133375'
+
+        self._request_count = 0
+
         self._session, self._url, self._chunk_size = session, url, chunk_size
-        self._length = int(head.headers["Content-Length"])
+
+        # initial range request for the end of the file
+        try:
+            tail = self._stream_response(start="", end=CONTENT_CHUNK_SIZE)
+        except HTTPError as e:
+            if e.response.status_code != 416:
+                raise
+
+            # The 416 response message contains a Content-Range indicating an
+            # unsatisfied range (that is a '*') followed by a '/' and the current
+            # length of the resource. E.g. Content-Range: bytes */12777
+            content_length = int(e.response.headers["content-range"].rsplit("/", 1)[-1])
+            tail = self._stream_response(start=0, end=content_length)
+
+        # e.g. {'accept-ranges': 'bytes', 'content-length': '10240',
+        # 'content-range': 'bytes 12824-23063/23064', 'last-modified': 'Sat, 16
+        # Apr 2022 13:03:02 GMT', 'date': 'Thu, 21 Apr 2022 11:34:04 GMT'}
+
+        if tail.status_code != 206:
+            if (
+                tail.status_code == 200
+                and int(tail.headers["content-length"]) <= CONTENT_CHUNK_SIZE
+            ):
+                # small file
+                content_length = len(tail.content)
+                tail.headers["content-range"] = f"0-{content_length-1}/{content_length}"
+            else:
+                raise HTTPRangeRequestUnsupported("range request is not supported")
+
+        # lowercase content-range to support s3
+        self._length = int(tail.headers["content-range"].partition("/")[-1])
```

Comment:
I thought `headers` is a special case-insensitive mapping anyway, does lower-casing actually matter?

Reply:
We used a requests ->  boto3 adapter that is case sensitive.

Reply:
So this is an S3 issue? I’d very much prefer this weird context to not get into pip. It’s OK we use lower-cased keys here (because it does not really matter for us), but the comment confuses things.